### PR TITLE
PS-547 - Added Support for Trying Multiple Different Launch Templates

### PR DESCRIPTION
**Pull request by slink1993**
## Purpose

Issues:
https://wavetrak.atlassian.net/secure/RapidBoard.jspa?rapidView=120&modal=detail&selectedIssue=PS-547&assignee=5cf98b5fb06c540e8258102f
This PR adds support for multiple launch templates to be used by iterating through the list of launch templates if one fails.

## Approach
- Updated `launch_template` in `main.py` to `launch_templates` to support list type, as well as the `launch_template` member in the `TempInstance` class.
- Added a new function `__launch_instance`, that implements the following functionality:
    - Iterates through the `launch_templates` given. If launching an EC2 instance at the current index of the `launch_templates` list succeeds, then the EC2 Instance metadata is returned. 
    - Otherwise if launching an EC2 instance at the current index fails, then the function will retry with the next index of the `launch_templates`. If all `launch_templates` fail, then an exception is raised and the program exits.

## TODO

- [ ] Deploy new version to Docker Hub

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -16,15 +16,15 @@ def main():
 
     name = os.environ['NAME']
     key_name = f'{name}-{uuid4()}'
-    launch_template_name = os.environ['LAUNCH_TEMPLATE_NAME']
+    launch_template_names = os.environ['LAUNCH_TEMPLATE_NAMES'].split()
```
Comment (jawang35) [inline main.py]: I think we should split on `,`. Spaces can be difficult to work with in the CLI.

File: lib/ec2.py [hunk 2]
```diff
@@ -54,45 +56,60 @@ class TempInstance():
...
-                        },
-                    ],
-                },
-            ],
-        )[0]
-        logger.info(f'Waiting for instance {self.instance.instance_id} to be '
-                    'ready...')
+    def __launch_instance(self, index: int = 0):
+        if index == len(self.launch_template_names):
+            raise Exception('No instance can be launched.')
```
Comment (jawang35) [inline lib/ec2.py:78]: I think we should check this variable in `main` and raise the exception there.

File: lib/ec2.py [hunk 3]
```diff
@@ -54,45 +56,60 @@ class TempInstance():
...
+        self.instance.wait_until_running()
+        return self.instance
 
+    def __enter__(self):
         try:
-            self.instance.wait_until_running()
-            return self.instance
-        except BaseException:
+            return self.__launch_instance()
+        finally:
```
Comment (jawang35) [inline lib/ec2.py:116]: Cool glad this worked.

**Comment by jawang35** [comment]
@stephenlink1993 also we should update the `README.md`.

File: main.py (hunk 1 above)
Comment (slink1993) [reply main.py]: good idea, updated 👍

**Comment by slink1993** [comment]
@jawang35 updated the `README.md`

File: README.md [hunk 4]
```diff
@@ -11,7 +11,10 @@ Docker image creates a Key Pair and initializes an EC2 instance both with the `N
 These variables can be [passed into the Docker run](https://docs.docker.com/engine/reference/commandline/run/#set-environment-variables--e---env---env-file) using the `-e` or `--env-file` flags:
 
 - `NAME` - Used to name EC2 Key Pair and Instance. UUID is appended to the end.
-- `LAUNCH_TEMPLATE_NAME` - Name of launch template to launch EC2 Instance.
+- `LAUNCH_TEMPLATE_NAMES` - A list of launch templates to be used to create EC2 instances. The
```
Comment (jawang35) [inline README.md]: `A comma separated list...` might be clearer.

File: README.md [hunk 5]
```diff
@@ -11,10 +11,7 @@ Docker image creates a Key Pair and initializes an EC2 instance both with the `N
 These variables can be [passed into the Docker run](https://docs.docker.com/engine/reference/commandline/run/#set-environment-variables--e---env---env-file) using the `-e` or `--env-file` flags:
 
 - `NAME` - Used to name EC2 Key Pair and Instance. UUID is appended to the end.
-- `LAUNCH_TEMPLATE_NAMES` - A list of launch templates to be used to create EC2 instances. The
-launch templates will be used in the order they are passed in. If creating an EC2 instance
-with a launch template fails, then the next launch template given in the list will be used.
-If creating an EC2 instance with every launch templates fails, then the program will exit.
+- `LAUNCH_TEMPLATE_NAMES` - A comma separated list of launch templates to be used to create EC2 instances (Ex: `launch-template-1, launch-template-2`). The launch templates will be used in the order they are passed in. If creating an EC2 instance with a launch template fails, then the next launch template given in the list will be used. If creating an EC2 instance with every launch templates fails, then the program will exit.
```
Comment (jawang35) [inline README.md]: I think the example shouldn't have spaces or we should trim after splitting.

**Review by jawang35 — APPROVED** [review]

**Pull request merged**